### PR TITLE
sysrib Encap rule programming for BGP GUE

### PR DESCRIPTION
**Pull request by wenovus**
Does not include the reconciler logic. Will be added in following PR.

### CODE REVIEW

File: sysrib/server.go [hunk 1]
```diff
@@ -521,15 +589,18 @@ func convertZebraRoute(niName string, zroute *zebra.IPRouteBody) *Route {
 			NetworkInstance: niName,
 		})
 	}
+	var routePref RoutePreference
+	switch zroute.Type {
+	case zebra.RouteBGP:
+		routePref.AdminDistance = 20
```
Comment (DanG100) [inline sysrib/server.go:595]: should this be configureable? if so TODO, if not make it const

File: sysrib/server.go [hunk 2]
```diff
@@ -281,6 +294,54 @@ func prefixString(prefix *pb.Prefix) (string, error) {
...
+	}
+	udp := layers.UDP{
+		SrcPort: 0, // TODO(wenbli): Implement hashing for srcPort.
+		DstPort: layers.UDPPort(binary.BigEndian.Uint16(gueHeaders.dstPort[:])),
+		Length:  34,
+	}
+
+	buf := gopacket.NewSerializeBuffer()
+	opts := gopacket.SerializeOptions{}
+	if err := udp.SerializeTo(buf, opts); err != nil {
```
Comment (DanG100) [inline sysrib/server.go:321]: can simplify to gopacket.SerializeLayers(buf, opts, ip, upd)

File: sysrib/sysrib.go [hunk 3]
```diff
@@ -48,9 +54,71 @@ type NIRIB struct {
...
+	}
+	ok, policy := sr.GUEPolicies.FindDeepestTag(*addr)
+	if !ok {
+		return GUEHeaders{}, false, nil
+	}
+	ip := net.ParseIP(address)
+	if ip == nil {
+		return GUEHeaders{}, false, fmt.Errorf("cannot parse IP address: %q", address)
+	}
+	switch policy.isV6 {
```
Comment (DanG100) [inline sysrib/sysrib.go:96]: change to if/else

File: sysrib/server.go [hunk 4]
```diff
@@ -281,6 +294,54 @@ func prefixString(prefix *pb.Prefix) (string, error) {
...
+func gueActions(gueHeaders GUEHeaders) ([]*fwdpb.ActionDesc, error) {
+	ip := layers.IPv4{
+		Version:  4,
+		IHL:      5,
+		Length:   74,
+		Protocol: layers.IPProtocolUDP,
+		SrcIP:    gueHeaders.srcIP4[:],
+		DstIP:    gueHeaders.dstIP4[:],
+	}
+	if gueHeaders.isV6 {
```
Comment (DanG100) [inline sysrib/server.go:308]: do you need to make layers.IPv6 header instead?

File: sysrib/sysrib.go [hunk 5]
```diff
@@ -48,9 +54,71 @@ type NIRIB struct {
...
 }
 
+// GUEPolicy represents the static values in the IP and UDP headers that are
+// to encapsulate the packet.
+//
+// srcPort is a hash.
+// dstIP is the nexthop of the BGP route.
+type GUEPolicy struct {
+	dstPort [2]byte
+	srcIP4  [4]byte
```
Comment (DanG100) [inline sysrib/sysrib.go:64]: i think this code would be simpler if you had one slice srcIP since you already is ivV6 bool

File: sysrib/sysrib.go [hunk 6]
```diff
@@ -48,9 +54,71 @@ type NIRIB struct {
 	IPV4 *generics_tree.TreeV4[*Route]
 }
 
+// GUEPolicy represents the static values in the IP and UDP headers that are
+// to encapsulate the packet.
+//
+// srcPort is a hash.
+// dstIP is the nexthop of the BGP route.
+type GUEPolicy struct {
+	dstPort [2]byte
```
Comment (DanG100) [inline sysrib/sysrib.go:63]: uint16?

File: sysrib/server.go (hunk 1 above)
Comment (wenovus) [reply sysrib/server.go:595]: Made const.

File: sysrib/sysrib.go (hunk 5 above)
Comment (wenovus) [reply sysrib/sysrib.go:64]: I wanted to use an array type here to keep the types tight.

File: sysrib/server.go [hunk 7]
```diff
@@ -281,6 +301,51 @@ func prefixString(prefix *pb.Prefix) (string, error) {
...
+		}
+	}
+	udp := &layers.UDP{
+		SrcPort: 0, // TODO(wenbli): Implement hashing for srcPort.
+		DstPort: layers.UDPPort(gueHeaders.dstPort),
+		Length:  34,
+	}
+
+	buf := gopacket.NewSerializeBuffer()
+	gopacket.SerializeLayers(buf, gopacket.SerializeOptions{}, ip, udp)
```
Comment (DanG100) [inline sysrib/server.go:330]: this func returns an error

File: sysrib/sysrib.go (hunk 5 above)
Comment (DanG100) [reply sysrib/sysrib.go:64]: i'm not sure i understand i understand what advantage this has considering you check the boolean every time these are used. you also wouldn't convert to and from slice/array every time you use these

Comment (wenovus) [reply sysrib/sysrib.go:64]: Oh I remember now, this type has to be hashable in order to do the diffing of the resolved nexthops, so a slice, or net.IP, wouldn't work.

**Review by DanG100 — APPROVED** [review]

**Pull request merged**